crates/mun_codegen/src/code_gen/symbols.rs

```diff
@@ -1,44 +1,64 @@
 use super::abi_types::{gen_abi_types, AbiTypes};
-use crate::ir::dispatch_table::{DispatchTable, DispatchableFunction};
-use crate::ir::function;
-use crate::type_info::TypeInfo;
+use crate::ir::{
+    dispatch_table::{DispatchTable, DispatchableFunction},
+    function,
+};
+use crate::type_info::{TypeGroup, TypeInfo};
 use crate::values::{BasicValue, GlobalValue};
 use crate::IrDatabase;
-use hir::{Ty, TypeCtor};
+use hir::Ty;
 use inkwell::{
     attributes::Attribute,
     module::{Linkage, Module},
-    types::StructType,
+    targets::TargetMachine,
     values::{FunctionValue, IntValue, PointerValue, StructValue, UnnamedAddress},
     AddressSpace,
 };
-use std::collections::HashMap;
-
-pub fn type_info_query(db: &impl IrDatabase, ty: Ty) -> TypeInfo {
-    match ty {
-        Ty::Apply(ctor) => match ctor.ctor {
-            TypeCtor::Float => TypeInfo::from_name("@core::float"),
-            TypeCtor::Int => TypeInfo::from_name("@core::int"),
-            TypeCtor::Bool => TypeInfo::from_name("@core::bool"),
-            TypeCtor::Struct(s) => TypeInfo::from_name(s.name(db).to_string()),
-            _ => unreachable!("{:?} unhandled", ctor),
-        },
-        _ => unreachable!(),
-    }
-}
+use std::collections::{HashMap, HashSet};
 
 /// Construct an IR `MunTypeInfo` struct value for the specified `TypeInfo`
-fn type_info_ir(ty: &TypeInfo, module: &Module) -> StructValue {
-    let context = module.get_context();
-    let guid_values: [IntValue; 16] =
-        array_init::array_init(|i| context.i8_type().const_int(u64::from(ty.guid[i]), false));
-    context.const_struct(
-        &[
-            context.i8_type().const_array(&guid_values).into(),
-            intern_string(module, &ty.name).into(),
-        ],
-        false,
-    )
+fn type_info_ir<D: IrDatabase>(
+    db: &D,
+    target: &TargetMachine,
+    module: &Module,
+    types: &AbiTypes,
+    type_values: &mut HashMap<TypeInfo, PointerValue>,
```

Comment:
Maybe rename to `type_info_to_ptr` or something to make it a little clearer

Reply:
How about `global_type_info_map`?